File: apps/api/app/schemas/contestacao.py
from datetime import date, datetime
from typing import Literal, Optional

from pydantic import BaseModel, ConfigDict, Field, field_validator


class ContestacaoGerar(BaseModel):
    cliente_id: int
    processamento_id: int
    periodo_inicio: Optional[date] = None
    periodo_fim: Optional[date] = None

    @field_validator("periodo_fim")
    @classmethod
    def fim_apos_inicio(cls, v: Optional[date], info) -> Optional[date]:
        inicio = info.data.get("periodo_inicio")
        if v is not None and inicio is not None and v < inicio:
            raise ValueError("periodo_fim deve ser >= periodo_inicio")
        return v


class ContestacaoStatusUpdate(BaseModel):
    status: Literal["rascunho", "enviada", "em_analise", "deferida", "indeferida"]


class ContestacaoSaveEdit(BaseModel):
    html_content: str = Field(..., max_length=500_000)


class ContestacaoResponse(BaseModel):
    id: str
    cliente_id: int
    processamento_id: Optional[int]
    adquirente: str
    periodo_inicio: date
    periodo_fim: date
    valor_excesso_total: float
    status: str
    html_carta: Optional[str] = None
    created_by: str
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)

File: apps/web/src/lib/api/contestacao.ts
import { apiClient } from './client';

export type ContestacaoStatus = 'rascunho' | 'enviada' | 'em_analise' | 'deferida' | 'indeferida';

export interface ContestacaoResumoItem {
  id: string;
  status: ContestacaoStatus | string;
  adquirente: string;
  processamento_id: number | null;
  periodo_inicio: string | null;
  periodo_fim: string | null;
  valor_excesso_total: number;
  created_at: string | null;
  updated_at: string | null;
}

export interface ContestacoesPorCliente {
  cliente_id: number;
  nome: string;
  total: number;
  contestacoes: ContestacaoResumoItem[];
}

export interface ContestacaoResponse {
  id: string;
  cliente_id: number;
  processamento_id: number | null;
  adquirente: string;
  periodo_inicio: string;
  periodo_fim: string;
  valor_excesso_total: number;
  status: ContestacaoStatus;
  html_carta: string | null;
  created_by: string;
  created_at: string;
  updated_at: string;
}

export const contestacaoApi = {
  gerar: (cliente_id: number, processamento_id: number): Promise<{ contestacao_id: string }> =>
    apiClient.post('/contestacoes/gerar', { cliente_id, processamento_id }).then((r) => r.data),

  listar: (cliente_id?: number, status?: string): Promise<ContestacaoResponse[]> =>
    apiClient
      .get('/contestacoes', {
        params: {
          ...(cliente_id !== undefined ? { cliente_id } : {}),
          ...(status ? { status } : {}),
        },
      })
      .then((r) => r.data),

  obter: (id: string): Promise<ContestacaoResponse> =>
    apiClient.get(`/contestacoes/${id}`).then((r) => r.data),

  atualizarStatus: (id: string, status: ContestacaoStatus): Promise<ContestacaoResponse> =>
    apiClient.put(`/contestacoes/${id}/status`, { status }).then((r) => r.data),

  saveEdit: (id: string, html_content: string): Promise<{ saved: boolean }> =>
    apiClient.post(`/contestacoes/${id}/save-edit`, { html_content }).then((r) => r.data),

  downloadUrl: (id: string): string =>
    `${apiClient.defaults.baseURL}/contestacoes/${id}/download`,

  remover: (id: string): Promise<void> =>
    apiClient.delete(`/contestacoes/${id}`).then(() => undefined),

  porCliente: (clienteId: number): Promise<ContestacoesPorCliente> =>
    apiClient.get<ContestacoesPorCliente>(`/contestacoes/cliente/${clienteId}`).then((r) => r.data),
};
